### PR TITLE
SB-734 Implement a cron task for removal of timestamped Maven snapshot artifacts 

### PR DESCRIPTION
**Pull request by kate-novik**

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    public MavenSnapshotManager()
+    {
+    }
+
+    public void deleteTimestampedSnapshotArtifacts(Repository repository,
+                                                   String artifactPath,
+                                                   VersionCollectionRequest request,
+                                                   int numberToKeep,
+                                                   int keepPeriod)
+            throws IOException, ProviderImplementationException,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Could you please put all exceptions on a separate line, and do the same for all multi-line method parameters across the pull? Thanks! :)

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                   NoSuchAlgorithmException, ParseException,
+                   XmlPullParserException
+    {
+        LayoutProvider layoutProvider = getLayoutProvider(repository, layoutProviderRegistry);
+        if (layoutProvider.containsPath(repository, artifactPath))
+        {
+            logger.debug("Removal of timestamped Maven snapshot artifact " + artifactPath +
+                         " in '" + repository.getStorage()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Cosmetic: Could be on the same line.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            Versioning versioning = request.getVersioning();
+            Artifact artifact = ArtifactUtils.convertPathToArtifact(artifactPath);
+
+            if (!versioning.getVersions()
+                           .isEmpty())
+            {
+                for (String version : versioning.getVersions())
+                {
+
+                    Path versionBasePath = Paths.get(request.getArtifactBasePath()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Cosmetic: Could be on the same line.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+         */
+        Map<Integer, String> mapToRemove = new HashMap<>();
+
+        metadata.getVersioning()
+                .getSnapshotVersions()
+                .forEach(e ->
+                         {
+
+                             if (e.getExtension()
+                                  .equals("jar"))
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Please, fix this.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+                    .filter(e -> ArtifactUtils.isArtifact(e.getPath()) && removingSnapshots.contains(e.getName()))
+                    .forEach(e ->
+                             {
+                                 try
+                                 {
+
+                                     layoutProvider.delete(storageId, repository.getId(), e.getPath(), true);
+
+                                     String artifactName = e.getName();
+                                     artifactName = artifactName.substring(0, artifactName.lastIndexOf("."));
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Please, fix.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+                                   if (count != snapshots.size() - numberToKeep)
+                                   {
+                                       mapToRemove.put(k, v);
+                                       count++;
+                                   }
+                               }));
+        }
+        else if (numberToKeep == 0 && keepPeriod != 0)
+        {
+            snapshots.forEach(((k, v) ->
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Please, fix.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.List;
+
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Kate Novik.
+ */
+public class ArtifactLocationRemoveTimestampedSnapshotOperation
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: I think a better name would be just `RemoveTimestampedSnapshotsOperation`.

Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: I think a better name would be just `RemoveTimestampedSnapshotsOperation`.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    private String previousPath;
+
+    private int numberToKeep;
+
+    private int keepPeriod;
+
+    private MavenSnapshotManager mavenSnapshotManager;
+
+
+    public ArtifactLocationRemoveTimestampedSnapshotOperation()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: You can ignore this in Sonar.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    public void execute(Path path)
+    {
+        File f = path.toAbsolutePath()
+                     .toFile();
+
+        String[] list = f.list(new PomFilenameFilter());
+        List<String> filePaths = list != null ? Arrays.asList(list) : new ArrayList<>();
+
+        String parentPath = path.getParent()
+                                .toAbsolutePath()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: You can ignore this in Sonar.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+                if (visitedVersionPaths.contains(f))
+                {
+                    return;
+                }
+            }
+
+            if (logger.isDebugEnabled())
+            {
+                // We're using System.out.println() here for clarity and due to the length of the lines
+                System.out.println(parentPath);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: You can ignore this in Sonar.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+            }
+
+            List<File> versionDirectories = getVersionDirectories(Paths.get(parentPath));
+            if (versionDirectories != null)
+            {
+                getVisitedRootPaths().put(parentPath, versionDirectories);
+
+                VersionCollector versionCollector = new VersionCollector();
+                VersionCollectionRequest request = versionCollector.collectVersions(path.getParent()
+                                                                                        .toAbsolutePath());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: You can ignore this in Sonar.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+                if (logger.isDebugEnabled())
+                {
+                    for (File directory : versionDirectories)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: You can ignore this in Sonar.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java [hunk 13]
```diff
@@ -2,9 +2,12 @@
 
 import org.carlspring.maven.commons.util.ArtifactUtils;
 import org.carlspring.strongbox.artifact.coordinates.ArtifactCoordinates;
+import org.carlspring.strongbox.artifact.locator.ArtifactDirectoryLocator;
+import org.carlspring.strongbox.artifact.locator.handlers.ArtifactLocationGenerateMavenChecksumOperation;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java]: Please, fix.

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java [hunk 14]
```diff
@@ -459,6 +468,37 @@ public void undeleteTrash()
...
+                                                   String repositoryId,
+                                                   String artifactPath,
+                                                   int numberToKeep,
+                                                   int keepPeriod)
+            throws IOException
+    {
+        Storage storage = getConfiguration().getStorage(storageId);
+        Repository repository = storage.getRepository(repositoryId);
+
+        if (repository.getPolicy()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java]: Cosmetic: Could be on one line.

**Comment by carlspring** [comment]
Looks good so far! :)

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java (hunk 5 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: ok, i did

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java (hunk 4 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Ok, did

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java (hunk 1 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: Ok

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java (hunk 2 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: ok

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java (hunk 3 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: ok

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java (hunk 7 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: ok

Comment (kate-novik) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/handlers/ArtifactLocationRemoveTimestampedSnapshotOperation.java]: ok

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java (hunk 14 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java]: ok

File: strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java (hunk 13 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-indexing/src/main/java/org/carlspring/strongbox/services/impl/ArtifactManagementServiceImpl.java]: ok

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java (hunk 6 above)
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: ok

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+                             {
+                                 String version = e.getVersion();
+                                 snapshots.put(Integer.parseInt(ArtifactUtils.getSnapshotBuildNumber(version)),
+                                               version);
+                             }
+                         });
+
+        if (numberToKeep != 0 && snapshots.size() > numberToKeep)
+        {
+            snapshots.forEach(((k, v) ->
```
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: ok

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+
+    @Inject
+    private LayoutProviderRegistry layoutProviderRegistry;
+    @Inject
+    protected StorageProviderRegistry storageProviderRegistry;
+    @Inject
+    private MavenMetadataManager mavenMetadataManager;
+
+
+    public MavenSnapshotManager()
```
Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: ignore

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+     * To get map of removing timestamped snapshots
+     *
+     * @param metadata     type Metadata
+     * @param numberToKeep type int
+     * @param keepPeriod   type int
+     * @return type Map<Integer, String>
+     * @throws IOException
+     * @throws XmlPullParserException
+     */
+    private Map<Integer, String> getRemovingSnapshots(Metadata metadata,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: I think this method should be called `getRemovableTimestampedSnapshots`, as it's more intuitive.

Comment (kate-novik) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/storage/snapshot/MavenSnapshotManager.java]: ok

**Comment by carlspring** [comment]
Thanks! :)

**Pull request merged**